graphene_sqlalchemy/tests/models.py

```diff
@@ -66,24 +77,60 @@ class Reporter(Base):
     favorite_article = relationship("Article", uselist=False)
 
     @hybrid_property
-    def hybrid_prop(self):
+    def hybrid_prop_untyped(self):
+        return self.first_name
+
+    @hybrid_property
+    def hybrid_prop_str(self) -> str:
         return self.first_name
 
+    @hybrid_property
```

Comment:
Tests were reverted and I implemented a single Test with new, separate models just for this.

Some nice discoveries:

- Comparing **very deeply** nested (List[List[List[x]]] or more) fails because graphene.List() allocations differ in memory for every subtype generated. But from a GraphQL type validation standpoint, it makes no sense to compare memory allocations like that. So I switched our type comparisons to `str(type) == str(expected_type)` and we are now able to test these (very much edge) cases. 
- Together with the stuff mentioned above, I also devised a simple way to have a loop compare those properties, but also have a sane way to retrieve the failed property (so that we don't lose any kind of dev experience), so that the failed properties are very easy to spot (Screenshot provided below - I think it looks nice).

```py
def test_sqlalchemy_hybrid_property_type_inference():
    class ShoppingCartItemType(SQLAlchemyObjectType):
        class Meta:
            model = ShoppingCartItem
            interfaces = (Node,)

    class ShoppingCartType(SQLAlchemyObjectType):
        class Meta:
            model = ShoppingCart
            interfaces = (Node,)

    hybrid_prop_expected_types: Dict[str, Union[Scalar, Structure]] = {
        # Basic types
        "hybrid_prop_str": String,
        "hybrid_prop_int": Int,
        "hybrid_prop_float": Float,
        "hybrid_prop_bool": Boolean,
        "hybrid_prop_decimal": String,  # Decimals should be serialized Strings
        "hybrid_prop_date": Date,
        "hybrid_prop_time": Time,
        "hybrid_prop_datetime": DateTime,
        # Lists and Nested Lists
        "hybrid_prop_list_int": List(Int),
        "hybrid_prop_list_date": List(Date),
        "hybrid_prop_nested_list_int": List(List(Int)),
        "hybrid_prop_first_shopping_cart_item": ShoppingCartItemType,
        "hybrid_prop_shopping_cart_item_list": List(Int)
    }

    assert sorted(list(ShoppingCartType._meta.fields.keys())) == sorted([
        # Columns
        "id",
        # Append Hybrid Properties from Above
        *hybrid_prop_expected_types.keys()
    ])

    for hybrid_prop_name, hybrid_prop_expected_return_type in hybrid_prop_expected_types.items():
        hybrid_prop_field = ShoppingCartType._meta.fields[hybrid_prop_name]

        # this is a simple way of showing the failed property name
        # instead of having to unroll the loop.
        assert (hybrid_prop_name, str(hybrid_prop_field.type)) \
               == (hybrid_prop_name, str(hybrid_prop_expected_return_type))
        assert hybrid_prop_field.description is None  # "doc" is ignored by hybrid property
```

This is how a failed property type comparison is shown:

![image](https://user-images.githubusercontent.com/5620032/166000901-2f2ce4de-d842-4b29-bcab-b95319fb6921.png)

Reply:
Suggestions implemented.